dna/__main__.py

```diff
@@ -807,22 +850,82 @@ def aggregate_result_scores(results_to_agg: typing.List[typing.Dict]):
+                    agg_score_sd = np.std(column, ddof=1).tolist()
+                    flat_agg_problem_scores[col_name] = agg_score_mean
+                    flat_agg_problem_scores[col_name.replace("mean", "std")] = agg_score_sd
+
+            if 'scores_by_dataset_id' in col_name:
                 column = flat_problem_scores_to_agg_df[col_name].values
                 if isinstance(column[0], collections.Iterable):
-                    agg_score = np.mean(np.stack(column, axis=0), axis=0).tolist()
-                    flat_agg_problem_scores[col_name] = agg_score
+                    # combine the each index from all lists into a tuple aka [1, 0], [2, 1], [3, 2] becomes [(1, 2, 3), (0, 1, 2)]
+                    total_distribution = list(zip(*column))
+                    flat_agg_problem_scores[col_name] = total_distribution
                 elif column[0] is not None:
-                    agg_score = np.mean(column).tolist()
-                    flat_agg_problem_scores[col_name] = agg_score
+                    flat_agg_problem_scores[col_name] = column
 
         agg_problem_scores = utils.inflate(flat_agg_problem_scores)
         agg_problem_scores['aggregated_ids'] = list(problem_scores_to_agg_df['run_id'])
         agg_scores.append(agg_problem_scores)
-
     return agg_scores
 
 
+def create_distribution_plots(agg_results: dict, output_dir: str):
+    scores_by_dataset = agg_results[0]["test_scores"]["scores_by_dataset_id"]
+    if len(list(scores_by_dataset.keys())) == 0:
+        print("No results, skipping")
+        return
+
+    metric_keys = scores_by_dataset[list(scores_by_dataset.keys())[0]].keys()
+    results_dict = {}
+    # aggregates metrics over all datasets
+    for metric_key in metric_keys:
+        new_metric_list = []
+        for index, dataset_name in enumerate(scores_by_dataset.keys()):
+            new_metric_list.append(scores_by_dataset[dataset_name][metric_key])
+        # non-iterable metrics are only non-iterable two layers deep now
+        if isinstance(new_metric_list[0][0], collections.Iterable):
+            new_metric_list = list(zip(*new_metric_list)) # now we have tuples of tuples for each metric
+            flattened_results = [list(sum(tupleOfTuples, ())) for tupleOfTuples in new_metric_list]
+        else:
+            flattened_results = list(itertools.chain(*new_metric_list)) # flatten all the arrays into one array
+        results_dict[metric_key] = flattened_results
+
+    if not os.path.isdir(output_dir):
+        os.makedirs(output_dir)
+
+    # plot aggregate scores in violin plot and save to `output_dir`
+    for metric_key in metric_keys:
+        if isinstance(results_dict[metric_key][0], collections.Iterable):
+            for index in [1, 25, 100, -1]:
+                name_of_index = "all" if index == 1 else str(index)
+                name_of_metric = metric_key.replace("_", " ")
+                ax = sns.violinplot(x=results_dict[metric_key][index])
```

Comment:
can we move the plotting code to dna/plot.py

Reply:
Done